### ISSUE TITLE
Marshal.GenerateGuidForType for types with true HasElementType returns Guid.Empty

### ISSUE THREAD

**Issue by hughbe** [dotnet/corefx#30940]
```cs
public static IEnumerable<object[]> GenerateGuidForType_Valid_TestData()
{
    yield return new object[] { typeof(int) };
    yield return new object[] { typeof(int).MakePointerType() };
    // Crashes with .NET Framework.
    if (!PlatformDetection.IsFullFramework)
    {
        yield return new object[] { typeof(int).MakeByRefType() };
    }
    yield return new object[] { typeof(string) };
    yield return new object[] { typeof(string[]) };

    yield return new object[] { typeof(NonGenericClass) };
    yield return new object[] { typeof(GenericClass<string>) };
    yield return new object[] { typeof(AbstractClass) };

    yield return new object[] { typeof(NonGenericStruct) };
    yield return new object[] { typeof(GenericStruct<string>) };

    yield return new object[] { typeof(NonGenericInterface) };
    yield return new object[] { typeof(GenericInterface<string>) };

    yield return new object[] { typeof(GenericClass<>) };
    // Crashes with .NET Framework.
    if (!PlatformDetection.IsFullFramework)
    {
        yield return new object[] { typeof(GenericClass<>).GetTypeInfo().GenericTypeParameters[0] };
    }

    yield return new object[] { typeof(ClassWithGuidAttribute) };
}

[Theory]
[MemberData(nameof(GenerateGuidForType_Valid_TestData))]
public void GenerateGuidForType_ValidType_ReturnsExpected(Type type)
{
    if (type.HasElementType)
    {
        if (PlatformDetection.IsNetCore)
        {
            Assert.Equal(Guid.Empty, type.GUID);
            Assert.Equal(type.GUID, Marshal.GenerateGuidForType(type));
        }
        else
        {
            Assert.NotEqual(type.GUID, Marshal.GenerateGuidForType(type));
        }
    }
    else
    {
        Assert.Equal(type.GUID, Marshal.GenerateGuidForType(type));
    }
}
```

**Comment by danmosemsft** [comment]
If the new behavior is arguably "better" then I suggest just disable the test for NETFX in the crashing cases.

**Comment by AaronRobinsonMSFT** [comment]
It is quite a bit better actually. In .NET Framework, the failure is actually an A/V because the cast of the `TypeHandle` to retrieve a `MethodTable` results in a `null` access. I actually don't think returning `Guid.Empty` is the right thing, rather it should throw a reasonable exception instead, but definitely shouldn't A/V like .NET Framework.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas Thanks for clarifying on the other related/closed bug about Framework.

@stephentoub @danmosemsft @jkotas Do any of you have thoughts about changing this behavior? When the user calls `Marshal.GenerateGuidForType()` on a type, they want a proper `Guid` or in my opinion an exception should be thrown to indicate one isn't possible. Returning `Guid.Empty` seems like the wrong thing because it is literally never what you want or expect and generally breaks the invariant of why the user would ask for a `Guid` for a type.

The flip side to this argument is that according to the documentation of [`Marshal.GenerateGuidForType()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.generateguidfortype?view=netframework-4.7#System_Runtime_InteropServices_Marshal_GenerateGuidForType_System_Type_) this function call is equivalent to calling `Type.GUID`.

All this boils down to either we should verify the behavior in coreclr is correct in all instances (e.g. `Marshal.GenerateGuidForType(typeof(int).MakeByRefType()) == typeof(int).MakeByRefType().GUID`) or make the calling of `Marshal.GenerateGuidForType()` on this type fail with an exception. Open to other opinions as well.

**Comment by jkotas** [comment]
Here are my thoughts on this subject:

`Type.Guid` and `Marshal.GenerateGuidForType` return the same Guid in .NET Framework (when GenerateGuidForType does not crash). The implementation of the two in .NET Framework is equivalent, except that `Marshal.GenerateGuidForType` has bugs that make it crash the runtime in some cases. The current .NET Core implementation of `Marshal.GenerateGuidForType` that just forwards to `Type.Guid` is fine.

The `Type.Guid` really only makes sense for types that work with COM interop. I think it is fine for it to return null guid for types that do not work with COM interop, like `ref int`. This property is legacy baggage. I do not think it is worth spending time on improving it - it is going to break as many things as it is going to fix.

The set of framework types that support COM interop in .NET Core is going to smaller than the set in .NET Framework. We have said that we are not going to support the "every other type in mscorlib supports COM" feature from .NET Framework v1.

The algorithm for auto-generating COM Guids for types that do not have one explicitly specified is pretty fragile. Look for `GetStringizedItfDef` and `GetStringizedTypeLibGuidForAssembly`. This affects what we are going to do with TlbExp. We may consider requiring the guid to be always explicitly specified for .NET Core flavor of tlbexp.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I have been digging into the TlbExp experience and found those two functions a while ago - they are indeed 'interesting', I think fragile is a bit generous 😸 If there are no concerns about returning `Guid.Empty` I will step through the code to update my understanding and if I don't see anything obviously dubious then I will simply close this issue and consider it 'by-design'.

**Comment by luqunl** [comment]
> We may consider requiring the guid to be always explicitly specified for .NET Core flavor of tlbexp.

Yes. I would love to add this restriction if we can. it will make analyzer work a little bit easy. But we can only add this restriction for COM not WinRT.

>  If there are no concerns about returning Guid.Empty

return Guid.Empty is better "hard exception".

also in long team, we can deprecated API Marshal.GenerateGuidForType. since Type.Guid is exactly same behavior as Marshal.GenerateGuidForType
